ELMOD1.py

```diff
@@ -0,0 +1,317 @@
+
+        Parameters:
+        - records (List[SeqRecord]): List of SeqRecord objects to write.
+        """
+        output_dir = "fastq_filtrator_results"
+        output_path = f"{output_dir}/{self.output_filename}.fastq"
+
+        os.makedirs(output_dir, exist_ok=True)
+
+        SeqIO.write(records, output_path, "fastq")
+        print(f"Filtered sequences written to {output_path}")
+
+class BiologicalSequence(ABC):
+    """
+    Abstract base class for biological sequences.
+    """
+
+    def __init__(self, sequence: str):
+        """
+        Initialize a BiologicalSequence instance.
+
+        Parameters:
+        - sequence (str): The biological sequence.
+
+        The sequence is automatically converted to uppercase.
+        """
+        self.sequence = sequence.upper()
+        self.validate_alphabet()
+
+    def __len__(self) -> int:
+        """
+        Return the length of the sequence.
+
+        Returns:
+        - int: Length of the sequence.
+        """
+        return len(self.sequence)
+    
+    def __getitem__(self, index: int) -> str:
+        """
+        Get the character at the specified index in the sequence.
+
+        Parameters:
+        - index (int): The index to retrieve.
+
+        Returns:
+        - str: The character at the specified index.
+        """
+        return self.sequence[index]
+    
+    def __str__(self) -> str:
+        """
+        Return the string representation of the sequence.
+
+        Returns:
+        - str: String representation of the sequence.
+        """
+        return self.sequence
+    
+    @abstractmethod
+    def is_valid_alphabet(self) -> bool:
+        """
+        Check if the sequence contains a valid alphabet.
+
+        Returns:
+        - bool: True if the alphabet is valid, False otherwise.
+        """
+        pass
+
+    def validate_alphabet(self) -> None:
+        """
+        Validate the alphabet of the sequence.
+
+        Raises:
+        - ValueError: If the alphabet is not valid.
+        """
+        if not self.is_valid_alphabet():
+            raise ValueError("Invalid alphabet for {}: {}".format(self.__class__.__name__, str(self)))
+
+class NucleicAcidSequence(BiologicalSequence):
+    """
+    Class representing nucleic acid sequences.
+    """
+    def complement(self) -> 'NucleicAcidSequence':
+        """
+        Return the complement of the sequence.
+
+        Returns:
+        - NucleicAcidSequence: Complement of the sequence.
+        """
+        self.validate_alphabet()
+        return self.__class__(''.join([self.complement_base(base) for base in self.sequence]))
+    
+    def gc_content(self) -> float:
+        """
+        Calculate the GC content of the sequence.
+
+        Returns:
+        - float: GC content of the sequence.
+        """
+        self.validate_alphabet()
+        gc_count = sum(1 for base in self.sequence if base in {'G', 'C'})
+        return gc_count / len(self)
```

Comment:
Мне кажется, такая запись будет более понятна

```suggestion
        self.validate_alphabet()
        gc_count = self.sequence.count('G') + self.sequence.count('C')
        return gc_count / len(self)
```

Reply:
Но так получается подсчет за 2 прогона. В общем получается:
- Вариант немного эффективнее
- Вариант немного понятнее

Какой лучше я тут сказать не могу, ахах, на ваш вкус)